test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightResourceSerializerTests.cs

```diff
@@ -0,0 +1,766 @@
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.OData.Edm;
+using Microsoft.OData.Evaluation;
+using Microsoft.OData.JsonLight;
+using Xunit;
+
+namespace Microsoft.OData.Tests.JsonLight
+{
+    public class ODataJsonLightResourceSerializerTests
+    {
+        private const string ServiceUri = "http://tempuri.org";
+        private EdmModel model;
+        private MemoryStream stream;
+        private ODataMessageWriterSettings messageWriterSettings;
+        private EdmEntityType productEntityType;
+        private EdmEntityType categoryEntityType;
+        private EdmEntitySet categoriesEntitySet;
+
+        public ODataJsonLightResourceSerializerTests()
+        {
+            this.model = new EdmModel();
+
+            this.categoryEntityType = new EdmEntityType("NS", "Category", /* baseType */ null, /* isAbstract */ false, /* isOpen */ true, /* hasStream */ true);
+            this.productEntityType = new EdmEntityType("NS", "Product");
+
+            this.categoryEntityType.AddKeys(this.categoryEntityType.AddStructuralProperty("Id", EdmPrimitiveTypeKind.Int32));
+            this.categoryEntityType.AddStructuralProperty("Name", EdmPrimitiveTypeKind.String);
+            this.categoryEntityType.AddStructuralProperty("BestSeller", new EdmEntityTypeReference(this.productEntityType, true));
+
+            this.productEntityType.AddKeys(this.productEntityType.AddStructuralProperty("Id", EdmPrimitiveTypeKind.Int32));
+            this.productEntityType.AddStructuralProperty("Name", EdmPrimitiveTypeKind.String);
+            this.productEntityType.AddStructuralProperty("Category", new EdmEntityTypeReference(this.categoryEntityType, true));
+
+            this.model.AddElement(this.categoryEntityType);
+            this.model.AddElement(this.productEntityType);
+
+            var entityContainer = new EdmEntityContainer("NS", "Container");
+            this.model.AddElement(entityContainer);
+
+            this.categoriesEntitySet = new EdmEntitySet(entityContainer, "Categories", this.categoryEntityType, true);
+
+
+            this.stream = new MemoryStream();
+            this.messageWriterSettings = new ODataMessageWriterSettings
+            {
+                EnableMessageStreamDisposal = false,
+                Version = ODataVersion.V4,
+                ShouldIncludeAnnotation = ODataUtils.CreateAnnotationFilter("*")
+            };
+
+            this.messageWriterSettings.SetServiceDocumentUri(new Uri(ServiceUri));
+        }
+
+        [Theory]
+        [InlineData(false, "Name", "{")]
+        [InlineData(true, null, "{\"@odata.type\":\"#Collection(NS.Category)\"")]
+        [InlineData(true, "Prop", "{\"Prop@odata.type\":\"#Collection(NS.Category)\"")]
+        public async Task WriteResourceSetStartMetadataPropertiesAsync_WritesResourceSetMetadataProperties(bool isUndeclared, string propertyName, string expected)
+        {
+            var result = await SetupJsonLightResourceSerializerAndRunTestAsync(

```

Comment:
Lets use concrete types so in future this can catch any changes to the public api.

Reply:
The `SetupJsonLightResourceSerializerAndRunTestAsync` is a private method within this test class. It sets up an ODataJsonLightResourceSerializer, then runs the given test code asynchronously, and then flushes and reads the stream back as a `string` for assertion. `var` felt natural to use here rather than `string`